with version 0.5.1+git.1550869862.4283c564-10.1 (from yesterday) can't boot a VM with a non-base image volume

##  System Information

### Linux distribution

```Debian testing```

### Terraform version

```sh
terraform -v

Terraform v0.11.13
```

### Provider and libvirt versions

```sh
terraform-provider-libvirt -version

terraform-provider-libvirt 0.5.1
Compiled against library: libvirt 4.0.0
Using library: libvirt 5.0.0
Running hypervisor: QEMU 3.1.0
Running against daemon: 5.0.0
```

## Description of Issue/Question

I have a simple setup with:

- one base volume created from a local file image
- one volume based on the previous one
- one domain which disk is the previous one

When using version 0.5.1 it works like expected. Yesterday I installed 0.5.1+git.1550869862.4283c564-10.1 from binary, and now when applying terraform plan, there is no error but my domain can't boot on its disk with the error "could not read the boot disk".
Without changes in the tf files.

### Setup

(Please provide the full _main.tf_ file for reproducing the issue (Be sure to remove sensitive information)

```
provider "libvirt" {
        uri = "qemu+ssh://user@kvm.example.com/system"
}

resource "libvirt_volume" "debian_stretch_image" {
  name = "debian_stretch_image"
  source = "/tmp/debian-stretch-qemu"
  format = "qcow2"
}

resource "libvirt_volume" "debian_stretch_volume-bastion" {
  name = "debian_stretch_volume-bastion"
  base_volume_id = "${libvirt_volume.debian_stretch_image.id}"
}

resource "libvirt_domain" "terraform_bastion" {
  name = "terraform_bastion"
  vcpu = 1
  memory = 1024
  autostart = true
  disk {
    volume_id = "${libvirt_volume.debian_stretch_volume-bastion.id}"
  }
  network_interface {
    network_name = "default"
    hostname = "bastion"
  }
  graphics {
    type = "vnc"
    listen_type = "address"
  }
}
```

### Steps to Reproduce Issue

```
$ terraform apply
```

Then look at the state of the machine (I use virt-manager for that), I see in the console "Booting from Hard Disk... Boot failed: could not read the boot disk"

I have an improved version of the testcase and a reworked version of the code that fixes the issue. I will do a PR later, as I still need to understand how the bug was introduced.

```console
=== RUN   TestAccLibvirtVolume_BackingStoreTestByName
--- FAIL: TestAccLibvirtVolume_BackingStoreTestByName (20.11s)
    testing.go:538: Step 0 error: Check failed: Check 3/3 error: libvirt_volume.9adhdysj7t: Attribute 'size' expected "1073741824", got "1024"
FAIL
FAIL    github.com/dmacvicar/terraform-provider-libvirt/libvirt 20.145s
```

--- a/libvirt/resource_libvirt_volume.go
+++ b/libvirt/resource_libvirt_volume.go
@@ -164,13 +164,16 @@ func resourceLibvirtVolumeCreate(d *schema.ResourceData, meta interface{}) error
 		volumeDef.Capacity.Unit = "B"
 		volumeDef.Capacity.Value = size
 	} else {
+		// the volume does not have a source image to upload
 
-		// the volume does not have a source image to upload, first handle
-		// whether it has a backing image
-		//
+		// if size is given, set it to the specified value
+		if _, ok := d.GetOk("size"); ok {
+			volumeDef.Capacity.Value = uint64(d.Get("size").(int))
+		}
+
+		//first handle whether it has a backing image
 		// backing images can be specified by either (id), or by (name, pool)
 		var baseVolume *libvirt.StorageVol
-
 		if baseVolumeID, ok := d.GetOk("base_volume_id"); ok {
 			if _, ok := d.GetOk("base_volume_name"); ok {
 				return fmt.Errorf("'base_volume_name' can't be specified when also 'base_volume_id' is given")
@@ -179,9 +182,7 @@ func resourceLibvirtVolumeCreate(d *schema.ResourceData, meta interface{}) error
 			if err != nil {
 				return fmt.Errorf("Can't retrieve volume ID '%s': %v", baseVolumeID.(string), err)
 			}
-		}
-
-		if baseVolumeName, ok := d.GetOk("base_volume_name"); ok {
+		} else if baseVolumeName, ok := d.GetOk("base_volume_name"); ok {
 			baseVolumePool := pool
 			if _, ok := d.GetOk("base_volume_pool"); ok {
 				baseVolumePoolName := d.Get("base_volume_pool").(string)
@@ -202,24 +203,26 @@ func resourceLibvirtVolumeCreate(d *schema.ResourceData, meta interface{}) error
 			if err != nil {
 				return fmt.Errorf("Could not retrieve backing store definition: %s", err.Error())
 			}
-			if _, ok := d.GetOk("size"); ok {
-				// does the backing store have some size information?, check at least that it is not smaller than the backing store
-				volumeDef.Capacity.Value = uint64(d.Get("size").(int))
-				backingStoreVolumeDef, err := newDefVolumeFromLibvirt(baseVolume)
-				if err != nil {
-					return err
-				}
 
-				if backingStoreVolumeDef.Capacity != nil && volumeDef.Capacity.Value < backingStoreVolumeDef.Capacity.Value {
-					return fmt.Errorf("When 'size' is specified, it shouldn't be smaller than the backing store specified with 'base_volume_id' or 'base_volume_name/base_volume_pool'")
-				}
+			backingStoreVolumeDef, err := newDefVolumeFromLibvirt(baseVolume)
+			if err != nil {
+				return err
+			}
+
+			// if the volumen does not specify size, set it to the size of the backing store
+			if _, ok := d.GetOk("size"); !ok {
+				volumeDef.Capacity.Value = backingStoreVolumeDef.Capacity.Value
+			}
+
+			// Always check that the size, specified or taken from the backing store
+			// is at least the size of the backing store itself
+			if backingStoreVolumeDef.Capacity != nil && volumeDef.Capacity.Value < backingStoreVolumeDef.Capacity.Value {
+				return fmt.Errorf("When 'size' is specified, it shouldn't be smaller than the backing store specified with 'base_volume_id' or 'base_volume_name/base_volume_pool'")
 			}
 			volumeDef.BackingStore = &backingStoreFragmentDef
 		}
 	}
-	if _, ok := d.GetOk("size"); ok {
-		volumeDef.Capacity.Value = uint64(d.Get("size").(int))
-	}
+
 	data, err := xmlMarshallIndented(volumeDef)
 	if err != nil {
 		return fmt.Errorf("Error serializing libvirt volume: %s", err)

--- a/libvirt/resource_libvirt_volume.go
+++ b/libvirt/resource_libvirt_volume.go
@@ -209,7 +209,7 @@ func resourceLibvirtVolumeCreate(d *schema.ResourceData, meta interface{}) error
 				return err
 			}
 
-			// if the volumen does not specify size, set it to the size of the backing store
+			// if the volume does not specify size, set it to the size of the backing store
 			if _, ok := d.GetOk("size"); !ok {
 				volumeDef.Capacity.Value = backingStoreVolumeDef.Capacity.Value
 			}

--- a/libvirt/resource_libvirt_volume.go
+++ b/libvirt/resource_libvirt_volume.go
@@ -196,8 +196,9 @@ func resourceLibvirtVolumeCreate(d *schema.ResourceData, meta interface{}) error
 				return fmt.Errorf("Can't retrieve base volume with name '%s': %v", baseVolumeName.(string), err)
 			}
 		}
+
 		if baseVolume != nil {
-			backingStoreDef, err := newDefBackingStoreFromLibvirt(baseVolume)
+			backingStoreFragmentDef, err := newDefBackingStoreFromLibvirt(baseVolume)
 			if err != nil {
 				return fmt.Errorf("Could not retrieve backing store definition: %s", err.Error())
 			}
@@ -213,7 +214,7 @@ func resourceLibvirtVolumeCreate(d *schema.ResourceData, meta interface{}) error
 					return fmt.Errorf("When 'size' is specified, it shouldn't be smaller than the backing store specified with 'base_volume_id' or 'base_volume_name/base_volume_pool'")
 				}
 			}
-			volumeDef.BackingStore = &backingStoreDef
+			volumeDef.BackingStore = &backingStoreFragmentDef
 		}
 	}
 	if _, ok := d.GetOk("size"); ok {

--- a/libvirt/resource_libvirt_volume_test.go
+++ b/libvirt/resource_libvirt_volume_test.go
@@ -142,6 +142,8 @@ func TestAccLibvirtVolume_BackingStoreTestByID(t *testing.T) {
 				Check: resource.ComposeTestCheckFunc(
 					testAccCheckLibvirtVolumeExists("libvirt_volume.backing-" + random, &volume),
 					testAccCheckLibvirtVolumeIsBackingStore("libvirt_volume." + random, &volume2),
+					resource.TestCheckResourceAttr(
+						"libvirt_volume."+random, "size", "1073741824"),
 				),
 			},
 		},
@@ -171,6 +173,8 @@ func TestAccLibvirtVolume_BackingStoreTestByName(t *testing.T) {
 				Check: resource.ComposeTestCheckFunc(
 					testAccCheckLibvirtVolumeExists("libvirt_volume.backing-"+random, &volume),
 					testAccCheckLibvirtVolumeIsBackingStore("libvirt_volume." + random, &volume2),
+					resource.TestCheckResourceAttr(
+						"libvirt_volume."+random, "size", "1073741824"),
 				),
 			},
 		},

--- a/libvirt/resource_libvirt_volume_test.go
+++ b/libvirt/resource_libvirt_volume_test.go
@@ -122,8 +122,7 @@ func TestAccLibvirtVolume_Basic(t *testing.T) {
 func TestAccLibvirtVolume_BackingStoreTestByID(t *testing.T) {
 	var volume libvirt.StorageVol
 	var volume2 libvirt.StorageVol
-	randomVolumeResource := acctest.RandString(10)
-	randomVolumeName := acctest.RandString(10)
+	random := acctest.RandString(10)
 	resource.Test(t, resource.TestCase{
 		PreCheck:     func() { testAccPreCheck(t) },
 		Providers:    testAccProviders,
@@ -131,18 +130,18 @@ func TestAccLibvirtVolume_BackingStoreTestByID(t *testing.T) {
 		Steps: []resource.TestStep{
 			{
 				Config: fmt.Sprintf(`
-				resource "libvirt_volume" "%s" {
-					name = "%s"
+				resource "libvirt_volume" "backing-%s" {
+					name = "backing-%s"
 					size =  1073741824
 				}
-				resource "libvirt_volume" "backing-store" {
-					name = "backing-store"
-					base_volume_id = "${libvirt_volume.%s.id}"
+				resource "libvirt_volume" "%s" {
+					name = "%s"
+					base_volume_id = "${libvirt_volume.backing-%s.id}"
 			        }
-				`, randomVolumeResource, randomVolumeName, randomVolumeResource),
+				`, random, random, random, random, random),
 				Check: resource.ComposeTestCheckFunc(
-					testAccCheckLibvirtVolumeExists("libvirt_volume."+randomVolumeResource, &volume),
-					testAccCheckLibvirtVolumeIsBackingStore("libvirt_volume.backing-store", &volume2),
+					testAccCheckLibvirtVolumeExists("libvirt_volume.backing-" + random, &volume),
+					testAccCheckLibvirtVolumeIsBackingStore("libvirt_volume." + random, &volume2),
 				),
 			},
 		},
@@ -152,8 +151,7 @@ func TestAccLibvirtVolume_BackingStoreTestByID(t *testing.T) {
 func TestAccLibvirtVolume_BackingStoreTestByName(t *testing.T) {
 	var volume libvirt.StorageVol
 	var volume2 libvirt.StorageVol
-	randomVolumeResource := acctest.RandString(10)
-	randomVolumeName := acctest.RandString(10)
+	random := acctest.RandString(10)
 	resource.Test(t, resource.TestCase{
 		PreCheck:     func() { testAccPreCheck(t) },
 		Providers:    testAccProviders,
@@ -161,17 +159,18 @@ func TestAccLibvirtVolume_BackingStoreTestByName(t *testing.T) {
 		Steps: []resource.TestStep{
 			{
 				Config: fmt.Sprintf(`
-					resource "libvirt_volume" "%s" {
-						name = "%s"
-						size =  1073741824
-					}
-					resource "libvirt_volume" "backing-store" {
-						name = "backing-store"
-						base_volume_name = "${libvirt_volume.%s.name}"
-				  }	`, randomVolumeResource, randomVolumeName, randomVolumeResource),
+				resource "libvirt_volume" "backing-%s" {
+					name = "backing-%s"
+					size =  1073741824
+				}
+				resource "libvirt_volume" "%s" {
+					name = "%s"
+                    base_volume_name = "${libvirt_volume.backing-%s.name}"
+			        }
+				`, random, random, random, random, random),
 				Check: resource.ComposeTestCheckFunc(
-					testAccCheckLibvirtVolumeExists("libvirt_volume."+randomVolumeResource, &volume),
-					testAccCheckLibvirtVolumeIsBackingStore("libvirt_volume.backing-store", &volume2),
+					testAccCheckLibvirtVolumeExists("libvirt_volume.backing-"+random, &volume),
+					testAccCheckLibvirtVolumeIsBackingStore("libvirt_volume." + random, &volume2),
 				),
 			},
 		},

